cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreImpl.java

```diff
@@ -1163,18 +1164,22 @@ public Object call(SQLDatabase db) throws Exception {
                         long docNumericId = getNumericIdInQueue(db, item.rev.getId());
                         long seq = 0;
 
+                        InsertRevisionCallable.Result result;
                         if (docNumericId != -1) {
-                            seq = doForceInsertExistingDocumentWithHistory(db, item.rev, docNumericId, item.revisionHistory,
+                             result = doForceInsertExistingDocumentWithHistory(db, item.rev, docNumericId, item.revisionHistory,
                                     item.attachments);
-                            item.rev.initialiseSequence(seq);
                             // TODO fetch the parent doc?
                             documentUpdated = new DocumentUpdated(null, item.rev);
                         } else {
-                            seq = doForceInsertNewDocumentWithHistory(db, item.rev, item.revisionHistory);
-                            item.rev.initialiseSequence(seq);
+                            result  = doForceInsertNewDocumentWithHistory(db, item.rev, item.revisionHistory);
                             documentCreated = new DocumentCreated(item.rev);
                         }
 
+                        item.rev.initialiseSequence(result.sequence);
```

Comment:
The events don't hold the seq directly, so using the value of reference semantics we can mutate the DocumentRevs before sending the events. And we don't send the events if its a duplicate, because the remaining part of the loop is skipped if there is a duplicate.